$.op() - wraps S.op() within S.bind() or S.bindAll()

rr326 opened the pull request:
This works.  

But I actually think the better solution is either to fix the bind/bindAll monkeypatching or don't monkeypatch and just create $.bind
## Fix monkeypatching

It think this is the best solution.  The monkeypatching, I think, is broken, since it doesn't just improve S.bind / S.bindAll, but actually breaks it.  For instance, without reslate, inside a bind you can do `$.op("resize", { "width" : "-10%", "height" : "+0"  })`. With reslate installed, that doesn't work. Instead you would have to do:

```
S.bind('x', function(win){
    var tmp=S.op("resize", { "width" : "-10%", "height" : "+0"  });
    tmp.run();
})
```

To fix this, the monkeypatching needs to recognize when the passed-in argument is a JSOperationWrapper object. If so, the monkeypatching would then automatically wrap as my $.op() function now does.  Unfortunately I couldn't figure out how to do it.
## Create new $.bind() / $.bindAll() functions

If the monkeypatching can't be fixed, then don't monkeypatch - just create new bind/bindAll functions within $. That way, S.bind() accepts standard slate S.op() functions as normal, and $.bind() is explicit that it only accepts $.xx operations (such as $.op()).

rr326 committed:
Added $.op() function.

This wraps a S.op() call so that it can be called from  within a reslate bind function.

rr326 committed:
(removed .idea from git)

rr326 committed:
added .gitignore

rr326 committed:
This is my working .slate.js

rr326 committed:
layout working

rr326 committed:
layouts working

lunixbochs commented:
> the monkeypatching needs to recognize when the passed-in argument is a JSOperationWrapper object

This is exactly the point of my X.run theory. It's a potential way to detect JSOperationWrapper objects (duck typing). I can also just monkeypatch S.op(). Either way I prefer it over duplicating more commands on $.

rr326 commented:
Ah - I think I get it.  You're saying that if you get something you don't
recognize, see if it has a .run property which is a function. If so, we can
be pretty confident is is a JSOperationWrapper, so just treat it as such
(and wrap and run)?

On Fri, Jun 13, 2014 at 2:25 PM, Ryan Hileman notifications@github.com
wrote:

> > the monkeypatching needs to recognize when the passed-in argument is a
> > JSOperationWrapper object
> 
> This is exactly the point of my X.run theory. It's a potential way to
> detect JSOperationWrapper objects (duck typing). I can also just
> monkeypatch S.op(). Either way I prefer it over duplicating more commands
> on $.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/lunixbochs/reslate/pull/11#issuecomment-46061630.

rr326 committed:
reslate.js
    - removed $.op()
    - Now bind monkeypatching checks to see if a JSOperationWrapper is passed in (_.isFunction(op.run)) and
        then wraps it accordingly
    - So now you can use S.op() calls directly

.slate.js
    - changed $.op to S.op
    - references to local directory (use sym links ot point to source)

rr326 committed:
removed reference to op

rr326 commented:
Your idea worked like a charm. Going to close this and open a new one.
